helper/helper_test.go

```diff
@@ -1,12 +1,185 @@
+
 	"github.com/greenplum-db/gpbackup/utils"
+	"golang.org/x/sys/unix"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+
+	"github.com/greenplum-db/gpbackup/toc"
+	"github.com/pkg/errors"
+)
+
+var (
+	testDir     = "/tmp/helper_test/20180101/20180101010101"
+	testTocFile = fmt.Sprintf("%s/test_toc.yaml", testDir)
 )
 
+type restoreReaderTestImpl struct {
+	waitCount int
+}
+
+func (r *restoreReaderTestImpl) waitForPlugin() error {
+	r.waitCount++
+	return nil
+}
+
+func (r *restoreReaderTestImpl) positionReader(pos uint64, oid int) error {
+	return nil
+}
+
+func (r *restoreReaderTestImpl) copyData(num int64) (int64, error) {
+	return 1, nil
+}
+
+func (r *restoreReaderTestImpl) copyAllData() (int64, error) {
+	return 1, nil
+}
+
+func (r *restoreReaderTestImpl) closeFileHandle() {
+}
+
+func (r *restoreReaderTestImpl) getReaderType() ReaderType {
+	return "nil"
+}
+
+type helperTestStep struct {
+	restorePipeWriterArgExpect string
+	restorePipeWriterResult    bool
+	skipFileArgTableOid        int
+	skipFileResult             bool
+}
+
+type restoreMockHelperImpl struct {
+	currentStep      int
+	started          bool
+	expectedOidBatch []oidWithBatch
+	expectedSteps    []helperTestStep
+
+	pipesMap    map[string]string // Ginkgo matcher works over map value, will diplicate key here
+	restoreData *restoreReaderTestImpl
+}
+
+func (h *restoreMockHelperImpl) openedPipes() []string {
+	if h.pipesMap == nil {
+		h.pipesMap = make(map[string]string)
+
+		for k := range pipesMap {
+			h.pipesMap[k] = k
+		}
+	}
+	ret := make([]string, 0, len(h.pipesMap))
+	for k := range h.pipesMap {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+func (h *restoreMockHelperImpl) makeStep() helperTestStep {
+	if !h.started {
+		h.started = true
+	} else {
+		h.currentStep++
+	}
+
+	Expect(h.currentStep).To(BeNumerically("<", len(h.expectedSteps)))
+	ret := h.expectedSteps[h.currentStep]
+	return ret
+}
+
+func (h *restoreMockHelperImpl) getCurStep() helperTestStep {
+	Expect(h.currentStep).To(BeNumerically("<", len(h.expectedSteps)))
+	return h.expectedSteps[h.currentStep]
+}
+
+func (h *restoreMockHelperImpl) closeAndDeletePipe(tableOid int, batchNum int) {
+}
+
+func newHelperTest(batches []oidWithBatch, steps []helperTestStep) *restoreMockHelperImpl {
+	var ret = new(restoreMockHelperImpl)
+	ret.expectedOidBatch = batches
+	ret.expectedSteps = steps
+	ret.pipesMap = nil
+	ret.restoreData = &restoreReaderTestImpl{}
+
+	return ret
+}
+
+func (h *restoreMockHelperImpl) getOidWithBatchListFromFile(oidFileName string) ([]oidWithBatch, error) {
+	return h.expectedOidBatch, nil
+}
+
+func (h *restoreMockHelperImpl) checkForSkipFile(pipeFile string, tableOid int) bool {
+	step := h.getCurStep()
+	Expect(tableOid).To(Equal(step.skipFileArgTableOid))
+	ret := step.skipFileResult
+	return ret
+}
+
+func (h *restoreMockHelperImpl) createPipe(pipe string) error {
+	// Check that pipe was not opened yet
+	Expect(h.openedPipes()).ShouldNot(ContainElement(pipe))
```

Comment:
Why do we need the `openedPipes` method? Why can't we just check `h.pipesMap`?

Reply:
h.pipesMap is a map, which is used as a set, to guarantee unique values. openedPipes() returns a slice, which is usable for checkers like ContainElement()

Reply:
hm,,,,why not to change 
```
pipesMap    map[string]string
```
to 
```
pipesMap    map[string]bool
```
?
In thic case we can change the checks like 
```
Expect(h.openedPipes()).ShouldNot(ContainElement(pipe))
```
to smth like 
```
Expect(h.pipesMap[pipe]).To(Equal(false))
```

and remove `openedPipes`

Reply:
One of the ideas of the ginkgo framework is to write and read test expectations more naturally on human's language. This is a good example of it, if omit braces then 
```
Expect openedPipes ShouldNot ContainElement pipe
```
is way more readable then
```
Expect pipesMap pipe To Equal false
```
Although I do agree that map[string]bool is enough here, and  map[string]string is not really needed.

Reply:
I suppose there is a way to change the name of `openedPipes()` so it would be readable enough and will not excessive actions like gathering of a slice of strings, when we already have a map to check the existence.

Reply:
I think it's possible if we wrap check into helper struct.